File: app/src/main/java/com/bopr/android/smailer/util/validator/TextValidator.java
package com.bopr.android.smailer.util.validator;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;
import android.widget.TextView;

import com.bopr.android.smailer.util.draw.WavyUnderlineSpan;

/**
 * Abstract {@link EditText} validator.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public abstract class TextValidator implements TextWatcher {

    private final TextView view;
    private Object span;

    public TextValidator(TextView view, int errorColor) {
        this.view = view;
        span = new WavyUnderlineSpan(errorColor);
    }

    /**
     * Returns true if input is valid.
     */
    public abstract boolean isValidInput(TextView textView, Editable editable, String text);

    @Override
    public void afterTextChanged(Editable editable) {
        if (isValidInput(view, editable, editable.toString())) {
            editable.removeSpan(span);
        } else {
            editable.setSpan(span, 0, editable.length(), 0);
        }
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count,
                                  int after) { /* do nothing */ }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) { /* do nothing */ }

}

File: app/src/main/kotlin/com/bopr/android/smailer/ui/TextValidator.kt
package com.bopr.android.smailer.ui

import android.text.Editable
import android.text.TextWatcher
import android.widget.TextView
import androidx.core.content.ContextCompat.getColor
import com.bopr.android.smailer.R

/**
 * Abstract [TextView] input validator.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
abstract class TextValidator(protected val view: TextView) : TextWatcher {

    private var underwaveSpan: WavyUnderlineSpan? = null

    /**
     * Returns true if text should be underlined with red wavy line.
     */
    protected open fun shouldUnderwave(text: String?): Boolean {
        return false
    }

    /**
     * Returns validation error message.
     */
    protected open fun getErrorMessage(text: String?): String? {
        return null
    }

    override fun afterTextChanged(editable: Editable) {
        val text = editable.toString()

        view.error = getErrorMessage(text)

        if (shouldUnderwave(text)) {
            if (underwaveSpan == null){
                underwaveSpan = WavyUnderlineSpan(getColor(view.context, R.color.errorLine))
            }
            editable.setSpan(underwaveSpan, 0, editable.length, 0)
        } else {
            editable.removeSpan(underwaveSpan)
        }
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        /* do nothing */
    }

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        /* do nothing */
    }
}